pallets/bridge_transfer/src/lib.rs

```diff
@@ -179,122 +265,93 @@ pub mod pallet {
-		}
 	}
 
 	impl<T: Config> MessageOriginInfo for Pallet<T> {
 		type Config = T;
 	}
 
 	impl<T: Config> Pallet<T> {
-		pub fn lottery_output(payload: &Lottery, dest_id: bridge::BridgeChainId) -> DispatchResult {
-			ensure!(
-				<bridge::Pallet<T>>::chain_whitelisted(dest_id),
-				Error::<T>::InvalidTransfer
-			);
-			let metadata: Vec<u8> = payload.encode();
-			<bridge::Pallet<T>>::transfer_generic(dest_id, T::BridgeLotteryId::get(), metadata)
+		pub fn asset_balance(asset: &bridge::ResourceId, who: &T::AccountId) -> BalanceOf<T> {
+			BridgeBalances::<T>::get(asset, who).unwrap_or(Zero::zero())
+		}
+
+		pub fn calculate_fee(dest_id: bridge::BridgeChainId, amount: BalanceOf<T>) -> BalanceOf<T> {
+			let (min_fee, fee_scale) = Self::bridge_fee(dest_id);
+			let fee_estimated = amount * fee_scale.into() / 1000u32.into();
+			if fee_estimated > min_fee {
+				fee_estimated
+			} else {
+				min_fee
+			}
 		}
 
-		pub fn on_message_received(message: DecodedMessage<Lottery>) -> DispatchResult {
-			// TODO.kevin: check the sender?
-			// Dest chain 0 is EVM chain, and 1 is ourself
-			Self::lottery_output(&message.payload, 0)
+		/// Deposit specific amount assets into recipient account.
+		///
+		/// If recipient is bridge account, assets would deposited to bridge account only,
+		/// otherwise, assets would be withdrawn from bridge account and then deposit to
+		/// recipient.
+		/// Bridge account is treat as holding account of all assets.
+		///
+		/// DO NOT guarantee asset was registered
+		/// DO NOT guarantee bridge account(e.g. hodling account) has enough balance
+		pub fn do_asset_deposit(
+			asset: &bridge::ResourceId,
+			recipient: &T::AccountId,
+			amount: BalanceOf<T>,
+		) {
+			let bridge_id = <bridge::Pallet<T>>::account_id();
+			let holding_balance = BridgeBalances::<T>::get(asset, &bridge_id).unwrap_or_default();
+			let recipient_balance = BridgeBalances::<T>::get(asset, recipient).unwrap_or_default();
+
+			if *recipient != bridge_id {
+				BridgeBalances::<T>::insert(
+					asset,
+					&bridge_id,
+					holding_balance.saturating_sub(amount),
+				);
+			}
+			BridgeBalances::<T>::insert(asset, recipient, recipient_balance.saturating_add(amount));
+		}
+
+		/// Withdraw specific amount assets from sender.
+		///
+		/// Assets would be withdrawn from the sender and then deposit to bridge account.
+		/// Bridge account is treat as holding account of all assets.
+		///
+		/// DO NOT guarantee asset was registered
+		/// DO NOT grarantee sender account has enough balance
+		pub fn do_asset_withdraw(
+			asset: &bridge::ResourceId,
+			sender: &T::AccountId,
+			amount: BalanceOf<T>,
+		) {
+			let bridge_id = <bridge::Pallet<T>>::account_id();
+			let holding_balance = BridgeBalances::<T>::get(asset, &bridge_id).unwrap_or_default();
+			let recipient_balance = BridgeBalances::<T>::get(asset, sender).unwrap_or_default();
+
+			BridgeBalances::<T>::insert(asset, sender, recipient_balance.saturating_sub(amount));
```

Comment:
If we return `Result`, we can use `checked_sub()` instead of `saturating_sub`. This can also guarantee there's always sufficient balance